File: lib/esparse/src/es2021/Finally.js
const Node = require('../Node');

// Finally[Yield, Await, Return]: {
//   `finally` Block[?Yield, ?Await, ?Return]
// }

let Finally = (Yield, Await, Return) => (given) => {
  let Block = require('./Block');
  let params = [Yield, Await, Return];
  let b = Block(Yield, Await, Return)(given);
  return new Node('Finally', given, 0, [b], params);
}

module.exports = Finally;

File: src/main/scala/kr/ac/kaist/jiset/js/ast/Finally.scala
package kr.ac.kaist.jiset.js.ast

import kr.ac.kaist.jiset.ir._
import kr.ac.kaist.jiset.util.Span
import kr.ac.kaist.jiset.util.Useful._
import io.circe._, io.circe.syntax._

trait Finally extends AST { val kind: String = "Finally" }

object AbsFinally extends Finally with AbsAST

object Finally {
  def apply(data: Json): Finally = AST(data) match {
    case Some(compressed) => Finally(compressed)
    case None => error("invalid AST data: $data")
  }
  def apply(data: AST.Compressed): Finally = {
    val AST.NormalCompressed(idx, subs, params, span) = data
    idx match {
      case 0 =>
        val x0 = subs(0).map(Block(_)).get
        Finally0(x0, params, span)
    }
  }
}

case class Finally0(x1: Block, parserParams: List[Boolean], span: Span) extends Finally {
  x1.parent = Some(this)
  def idx: Int = 0
  def k: Int = d(x1, 0)
  def fullList: List[(String, PureValue)] = l("Block", x1, Nil).reverse
  def maxK: Int = 0
  override def toString: String = {
    s"finally $x1"
  }
}
